Deck Picker Task Cancellation: ConcurrentModificationException

On my personal collection/device

Just after completing a deck, I was going through very quickly

[Anki-ConcurrentModificationLogcat.txt](https://github.com/ankidroid/Anki-Android/files/4887420/Anki-ConcurrentModificationLogcat.txt)


```
2020-07-07 23:22:07.143 12524-12524/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ichi2.anki, PID: 12524
    java.lang.RuntimeException: Unable to pause activity {com.ichi2.anki/com.ichi2.anki.DeckPicker}: java.util.ConcurrentModificationException
        at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4742)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4691)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:4626)
        at android.app.servertransaction.PauseActivityItem.execute(PauseActivityItem.java:45)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199)
        at android.os.Handler.dispatchMessage(Handler.java:112)
        at android.os.Looper.loop(Looper.java:216)
        at android.app.ActivityThread.main(ActivityThread.java:7625)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)
     Caused by: java.util.ConcurrentModificationException
        at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:967)
        at java.util.LinkedList$ListItr.next(LinkedList.java:889)
        at com.ichi2.async.CollectionTask.cancelAllTasks(CollectionTask.java:235)
        at com.ichi2.anki.DeckPicker.onPause(DeckPicker.java:867)
        at android.app.Activity.performPause(Activity.java:7663)
        at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1536)
        at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4726)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4691) 
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:4626) 
        at android.app.servertransaction.PauseActivityItem.execute(PauseActivityItem.java:45) 
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2199) 
        at android.os.Handler.dispatchMessage(Handler.java:112) 
        at android.os.Looper.loop(Looper.java:216) 
        at android.app.ActivityThread.main(ActivityThread.java:7625) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987) 
``

--- a/AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java
+++ b/AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java
@@ -135,7 +135,7 @@ public class CollectionTask extends BaseAsyncTask<TaskData, TaskData, TaskData> 
     /**
      * Tasks which are running or waiting to run.
      * */
-    private static List<CollectionTask> sTasks = Collections.synchronizedList(new LinkedList<>());
+    private static final List<CollectionTask> sTasks = Collections.synchronizedList(new LinkedList<>());
 
 
     /**
@@ -247,7 +247,7 @@ public class CollectionTask extends BaseAsyncTask<TaskData, TaskData, TaskData> 
                 return cancel(true);
             }
         } catch (Exception e) {
-            // Potentially catching SecurityEcexption, from
+            // Potentially catching SecurityException, from
             // Thread.interrupt from FutureTask.cancel from
             // AsyncTask.cancel
             Timber.w(e, "Exception cancelling task");
@@ -275,14 +275,13 @@ public class CollectionTask extends BaseAsyncTask<TaskData, TaskData, TaskData> 
     /** Cancel all tasks of type taskType*/
     public static void cancelAllTasks(TASK_TYPE taskType) {
         int count = 0;
-        synchronized (sTasks) {
-            for (CollectionTask task: sTasks) {
-                if (task.mType != taskType) {
-                    continue;
-                }
-                if (task.safeCancel()) {
-                    count ++;
-                }
+        // safeCancel modifies sTasks, so iterate over a concrete copy
+        for (CollectionTask task: new ArrayList<>(sTasks)) {
+            if (task.mType != taskType) {
+                continue;
+            }
+            if (task.safeCancel()) {
+                count++;
             }
         }
         if (count > 0) {

